Add protocol to URLs for thumbnails in cart.

peterfabian opened the pull request:
and removed `wc_get_relative_url()`, as it's not used anymore.

### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request, and the reason for such changes. -->

Closes #19819 .

This change replaces protocol-relative thumbnail URLs with protocol-specific URLs. It should resolve issues #19819 and #17577, mitigate problems introduced by using protocol-relative URLs such as [[1]](https://www.paulirish.com/2010/the-protocol-relative-url/) and [[2]](https://jeremywagner.me/blog/stop-using-the-protocol-relative-url/). 

Ad caching:
Cart hash is stored in a cookie which is served both via http and https, so any time a cart is updated, the widget cache gets updated both over http and https.
Since the sessionStoreage and localStorage are different for http and https, there should not be a problem with https cart page fetching images via http protocol and vice versa. Cart fragments are loaded via AJAX and they should be served correctly (unless AJAX call is cached, but then cart would not work anyway).

As single product page uses full URLs for images already, sites would most likely already be broken if they had images served from a CDN over http while running the checkout on https (links from checkout back to shop are https).
Even if a site is http only with checkout forcing https, then I see the following eventualities today:
- http pages: images will be served via //, thus fetched via http
- https pages (e.g. checkout): images served via //, thus fetched via https
Therefore, it looks to me that providing protocol in the beginning of the URL actually does not change much--even today the shop needs resources on both http and https if it wants to run in this mixed mode.

If there is a website that uses a more persistent widget cache (e.g. W3 Total Cache) that would cache the protocol-relative image URLs (i.e. starting with `//`), then if the widget is embedded/displayed on the checkout website accessed via https, using protocol-relative URLs means that the images would be accessed via https, while on other pages those would be accessed via http. Thus, the images should be present over both protocols already today if the site should work correctly with protocol-relative URLs.

I think the only problem happens when a resource is accessed via http, cached, then accessed via https protocol. To fix this behaviour, I slightly modified the widget id used for caching to include the scheme/protocol.

Please let me know if there are any other scenarios I might have missed, especially around CDNs serving over http only, I haven't tried that set up yet. (I guess if CDN serves over https, then there should not be any problem).


### How to test the changes in this Pull Request:
0. Set WC to not use https by default, only force it for checkout.
1. Add some products to cart.
2. Observe the images/thumbnails are displayed in cart.
3. Configure WC to run on a specific port, e.g. 8080
4. Observe the images are *not* displayed in cart.
5. Activate persistent object caching, e.g. W3 Total Cache with sufficiently large timeout.
6. Add `Top rated products` widget to the layout, e.g. on the shop site below header.
7. Refresh the page to cache the widget.
8. Go to the same page using https, observe mixed content warnings in the dev console, images missing from products in the widget (if they are blocked by default).
9. Apply branch. 
10. Refresh while accessing via http and https, check that both versions serve widget with correct links and images.
11. Go to cart.
12. Observe images are displayed correctly (in contrast to step 4).

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Replaced protocol-relative URLs for thumbnails in the cart and mini-cart with protocol-specific URLs to prevent problems with URLs that include port section.

peterfabian committed:
Add protocol to URLs for thumbnails in cart.

Removed wc_get_relative_url(), as it's not used anymore.

peterfabian committed:
Removed unused function wc_is_external_resource.

peterfabian committed:
Putting public functions back.

In case some plugins are using them.

peterfabian committed:
Removed superfluous newline.

peterfabian committed:
Add scheme to widget id when caching objects.

This should prevent cache from serving mixed content when widget cache is preloaded with http version, then accessed via https.

claudiulodro commented on includes/abstracts/abstract-wc-product.php:
```diff
@@ -1809,7 +1809,7 @@ public function get_image( $size = 'woocommerce_thumbnail', $attr = array(), $pl
 		}
 
-		return apply_filters( 'woocommerce_product_get_image', wc_get_relative_url( $image ), $this, $size, $attr, $placeholder, $image );
+		return apply_filters( 'woocommerce_product_get_image', $image, $this, $size, $attr, $placeholder );
```
We still need the last `$image` here, as removing that is a breaking change.

peterfabian replied:
Yup, so this has to wait for 4.0, if I get the SemVer right, correct? 
How do we flag such things for update when we can introduce breaking change (so that we won't forget to remove duplicate parameter)?

claudiulodro replied:
Yep you got it; we can't make breaking changes until 4.0. 👍 You can put a [todo](https://github.com/woocommerce/woocommerce/blob/094fc1bf3a25bd5e42becabbb219a7d9f52077ea/includes/admin/settings/class-wc-settings-checkout.php#L6) to flag it for removal in the next major release.

claudiulodro reviewed (changes requested):
This looks good. Just one small breaking change that needs attention.

peterfabian committed:
Added back extra parameter for backwards compatibility.

mikejolley reviewed (changes requested):
Good job 👍 Tested and works well. I also confirmed the issue I thought was an issue doesn't occur :)

Made some small suggestions to tweak, then this can be merged.

peterfabian committed:
Applied the filter inside the function.

Added optional scheme parameter and @since docstring according to the code review.

mikejolley committed:
Pass scheme without hyphen
